documentation for redaction configuration on listeners (#23568)
hashicorp/vault: website/content/docs/configuration/listener/index.mdx

# `listener` stanza

The `listener` stanza configures the addresses and ports on which Vault will
respond to requests. At this time, there are two listeners:
- [TCP][tcp]
- [Unix Domain Socket][unix]

[tcp]: /vault/docs/configuration/listener/tcp
[unix]: /vault/docs/configuration/listener/unix

## Unauthenticated endpoints - sensitive data redaction

There are four different types of information deemed sensitive that can be returned
by unauthenticated API endpoints:

1. Version number
2. Build date
3. Cluster name
4. IP address

Vault offers the ability to configure each `listener` stanza such that when appropriate,
these values will be redacted from responses.

The following API endpoints support redaction based on `listener` stanza configuration:

* [`/sys/health`](/vault/api-docs/system/health)
* [`/sys/leader`](/vault/api-docs/system/leader)
* [`/sys/seal-status`](/vault/api-docs/system/seal-status)

> When a value is redacted by Vault, it will be replaced with an empty string (`""`).
> 
> ~> Note: In certain situations, due to an empty string value, the related key may
> no longer be present in the response object returned from the API, as that field is
> omitted when the value is empty (`""`).
> 
> ~> Note: The Vault API is also consumed by the Vault CLI and UI, therefore enabling redaction
> settings will also affect them both.

Reviewer: ```suggestion
Vault replaces redacted information with an empty string (`""`). APIs that
omit response parameters with empty values also suppress redacted values in
response objects.

<Note title="Redacting values affects all API responses">
  The Vault CLI and UI consume Vault API responses. As a result, your redaction
  settings will apply to CLI and UI output in addition to direct API calls.
</Note>
```

Style correction: write in active voice, avoid back-to-back asides when possible

## Redaction examples

Please see [redaction settings](/vault/docs/configuration/listener#redaction-settings)

> (below) for details on each redaction setting.

Reviewer 2: 🤔 I wonder if it would make sense to provide the settings first so that the reader has further context prior to reading through the examples. What do you think?

Reviewer: +1, it's generally a better idea to provide instruction before the examples. Examples without context don't necessarily help the reader. Especially if the first sentence encourages the reader to jump to those instructions anyway 😅

Author: 😨 I missed this comment before I just hit merge! 😢 Lemme fix that.

Author: @ccapurso https://github.com/hashicorp/vault/pull/23572 👍🏼

> Example configuration for the [`tcp`](/vault/docs/configuration/listener/tcp) listener,
> enabling [`redact_addresses`](/vault/docs/configuration/listener#redact_addresses),
> [`redact_cluster_name`](/vault/docs/configuration/listener#redact_cluster_name) and
> [`redact_version`](/vault/docs/configuration/listener#redact_version).

Reviewer: In general, we should introduce examples with plain language that explains the outcome. In this case, we really don't need to explicitly list the relevant parameters since (1) the name makes it pretty obvious and (2) if we switch things around so the settings are first in the doc, in theory, the reader will already have an idea of what they're looking at and which parameters matter to the example :)  

```suggestion
Configure the [`tcp`](/vault/docs/configuration/listener/tcp) listener to
redact IP address, cluster name, and version information:
```

Style correction: write in the active voice, use declarative language to introduce instruction or examples.

```hcl

> ui            = true
> cluster_addr  = "https://127.0.0.1:8201"
> api_addr      = "https://127.0.0.1:8200"
> disable_mlock = true
> 
> storage "raft" {
>   path = "/path/to/raft/data"
>   node_id = "raft_node_1"
> }
> 
> listener "tcp" {
>   address             = "127.0.0.1:8200",
>   tls_cert_file = "/path/to/full-chain.pem"
>   tls_key_file  = "/path/to/private-key.pem"
>   redact_addresses    = "true"
>   redact_cluster_name = "true"
>   redact_version      = "true"
> }
> 
> telemetry {
>   statsite_address = "127.0.0.1:8125"
>   disable_hostname = true
> }

Reviewer: ```suggestion
listener "tcp" {
  address             = "127.0.0.1:8200",
  tls_cert_file = "/path/to/full-chain.pem"
  tls_key_file  = "/path/to/private-key.pem"
  redact_addresses    = "true"
  redact_cluster_name = "true"
  redact_version      = "true"
}
```

In general, we should focus on the most important information in example code. The stuff outside the `listener` stanza is irrelevant to the outcome described in the example description so it makes the example more accessible if we exclude it. Sometimes extra info is useful so folks can orient themselves, but in this case, the stanza is literally called `listener "tcp"` and it's a top-level configuration block, so the additional context isn't really needed.

```

> ### API: `/sys/health`

Reviewer: The headings here could create confusion because it makes it look like you're moving on to another example when, really, you're illustrating the results of the previous example. I would recommend streamlining this a bit:

1. Limit the examples to 2 calls: one that shows the redacted values returned as empty strings and one that shows the redacted values omitted from the response. 
2. Provide two example responses and label them as "without redaction" and "with redaction" so folks can see the difference in the responses for themselves.
3. Simplify the text about the differences to something like "In the unredacted response `field_1` is populated with `value_1`. In the redacted response, `field_1` is [empty | omitted]."

### API: `sys/leader`

`curl -s https://127.0.0.1:8200/v1/sys/leader | jq`:

```json
{
  "ha_enabled": true,
  "is_self": false,
  "active_time": "0001-01-01T00:00:00Z",
  "leader_address": "",
  "leader_cluster_address": "",
  "performance_standby": false,
  "performance_standby_last_remote_wal": 0,
  "raft_committed_index": 164,
  "raft_applied_index": 164
}
```

When comparing the JSON response above to when redaction settings are not enabled,
we can see that `leader_address` and `leader_cluster_address` are now redacted and
shown only as `""`.

### API: `sys/seal-status`

`curl -s https://127.0.0.1:8200/v1/sys/seal-status | jq`:

```json
{
  "type": "shamir",
  "initialized": true,
  "sealed": false,
  "t": 1,
  "n": 1,
  "progress": 0,
  "nonce": "",
  "version": "",
  "build_date": "",
  "migration": false,
  "cluster_id": "a1a7a078-0ae1-7fb9-41ec-2f4f583c773e",
  "recovery_seal": false,
  "storage_type": "raft"
}
```

When comparing the JSON response above to when redaction settings are not enabled,
we can see that `cluster_name` is not present (it is now omitted due to being empty).
Also, that `build_date` and `version` are now redacted and shown only as `""`.

> ### CLI: `vault status`

Reviewer: Similar feedback here. Having a heading here implies that the CLI example relates to configuring the listener, but it's really an illustration of what the response looks like. At minimum I would fold this into the discussion above either as a third example (if we feel the distinction between direct API and CLI calls is critical) or as one of the two examples mentioned previously.

```
Key                     Value
---                     -----
Seal Type               shamir
Initialized             true
Sealed                  false
Total Shares            5
Threshold               3
Version                 n/a
Build Date              n/a
Storage Type            raft
HA Enabled              true
HA Cluster              n/a
HA Mode                 standby
Active Node Address     <none>
Raft Committed Index    219
Raft Applied Index      219
```

When examining the output of `vault status` we can see that `Version`, `Build Date`

> and `HA Cluster` show as `n/a`. Also, that `Active Node Address` shows as `<none>`.

Reviewer: What does N/A vs none mean in this case? I'm assuming N/A means the API returned an empty string and "none" means the field was omitted from the response, but it would be helpful for folks if we made it clear which is which.

## Common configuration options

The following options apply to both types of listener (see above), and can be configured

> for each individual listener stanza.

Reviewer: I'm not sure this sentence adds much as folks typically assume a parameter applies unless we say otherwise. Something like "Use the following parameters to configure redaction for your listeners" would be more appropriate.

Author: I know what you mean, but I was trying to add something that meant the common config options section could be extended later if we had settings that weren't for just redaction (e.g. disabling an endpoint via config here).  🤔

> Please see the [sensitive data redaction](/vault/docs/configuration/listener#unauthenticated-endpoints-sensitive-data-redaction) explanation above for further information.

Reviewer: Including this link here doesn't make sense as we're expecting folks to have already skimmed past those sections to get here.  In general, we should avoid in-page links unless it's really needed from an accessibility or discoverability perspective.

```suggestion
```

### Redaction settings

> - `redact_addresses` `(bool: false)` - If enabled, will redact `leader_address` and `cluster_leader_address` values when applicable.

Reviewer: ```suggestion
- `redact_addresses` `(bool: false)` - Redacts `leader_address` and `cluster_leader_address` values in applicable API responses when `true`.
```

> - `redact_cluster_name` `(bool: false)` - If enabled, will redact `cluster_name` values when applicable.

Reviewer: ```suggestion
- `redact_cluster_name` `(bool: false)` - Redacts `cluster_name` values in applicable API responses when `true`.
```

> - `redact_version` `(bool: false)` - If enabled, will redact `version` and `build_date` values when applicable.

Reviewer: ```suggestion
- `redact_version` `(bool: false)` - Redacts `version` and `build_date` values in applicable API responses when `true`.
```
